src/shogun/mathematics/linalg/internal/implementation/Variance.h

```diff
@@ -0,0 +1,319 @@
+{
+	/** Scalar type */
+	typedef typename Vector::Scalar T;
+
+	/** Eigen vector type */
+	typedef Eigen::Matrix<T,Eigen::Dynamic,1> VectorXt;
+
+	/** Calculates unbiased empirical variance estimate of given values. Given
+	 * \f$\{x_1, ..., x_m\}\f$, this is
+	 * \f$\frac{1}{m-1}\sum_{i=1}^m (x_i-\bar{x})^2\f$ where
+	 * \f$\bar x=\frac{1}{m}\sum_{i=1}^m x_i\f$
+	 *
+	 * @param x vector of values
+	 * @return variance of given values
+	 */
+	static T compute(SGVector<T> x)
+	{
+		REQUIRE(x.vlen>1, "Please ensure that vector length is greater than 1.\n")
+
+		SGVector<T> squaredResult(x.vlen);
+		Eigen::Map<VectorXt> eigX = x;
+		Eigen::Map<VectorXt> eigSquaredResult = squaredResult;
+
+		T meanVal = mean<Backend::EIGEN3, SGVector<T>>::compute(x);
+		eigSquaredResult.fill(meanVal);
+		eigSquaredResult = eigX - eigSquaredResult;
+		eigSquaredResult = eigSquaredResult.array() * eigSquaredResult.array();
+
+		return ((T) 1 / (x.vlen - 1)) * eigSquaredResult.sum();
+	}
+};
+
+/**
+ * @Brief A generic class that computes the variance of a matrix column-wise.
+ */
+template<enum Backend, typename Matrix>
+struct colwise_variance
+{
+	/** Generic scalar type */
+	typedef typename Matrix::Scalar T;
+
+	/** Vector return type */
+	typedef SGVector<T> ReturnType;
+
+	/** Calculates unbiased empirical variance estimate of given values for each column of
+	 * an input matrix. Given a single column \f$k\f$ with entries \f$\{x_{1k}, ..., x_{mk}\}\f$ 
+	 * from matrix \f$x\f$, this is \f$\frac{1}{m-1}\sum_{i=1}^m (k_i-\bar{k})^2\f$ where
+	 * \f$\bar k=\frac{1}{m}\sum_{i=1}^m k_i\f$.  This is computed for each column \f$k\f$.
+	 *
+	 * Computes the variance for each column of a matrix
+	 *
+	 * @param x matrix of values
+	 * @return the column-wise variance of a matrix
+	 */
+	static ReturnType compute(Matrix x);
+};
+
+/**
+ * @Brief A specialization of colwise_variance that uses SGMatrix and SGVector as its types
+ * and uses Eigen3 as its backend component
+ */
+template<typename Matrix>
+struct colwise_variance<Backend::EIGEN3, Matrix>
+{
+
+	/** Generic scalar type */
+	typedef typename Matrix::Scalar T;
+
+	/** Vector return type */
+	typedef SGVector<T> ReturnType;
+
+	/** Eigen vector type */
+	typedef Eigen::Matrix<T,Eigen::Dynamic,1> VectorXt;
+
+	/** Calculates unbiased empirical variance estimate of given values for each column of
+	 * an input matrix. Given a single column \f$k\f$ with entries \f$\{x_{1k}, ..., x_{mk}\}\f$ 
+	 * from matrix \f$x\f$, this is \f$\frac{1}{m-1}\sum_{i=1}^m (k_i-\bar{k})^2\f$ where
+	 * \f$\bar k=\frac{1}{m}\sum_{i=1}^m k_i\f$.  This is computed for each column \f$k\f$.
+	 *
+	 * Computes the variance for each column of a matrix
+	 *
+	 * @param m matrix of values
+	 * @return the column-wise variance of a matrix
+	 */
+	static ReturnType compute(SGMatrix<T> m)
+	{
+		REQUIRE(m.num_rows > 0, "Please ensure that m has more than 0 rows.\n")
+		REQUIRE(m.num_cols > 0, "Please ensure that m has more than %d columns.\n")
+
```

Comment:
I had mentioned this above, but there currently is no implementation for a columnwise mean, so I'm leaving this unimplemented for now.